File: 0230-kth-smallest-element-in-a-bst/0230-kth-smallest-element-in-a-bst.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        stack = []
        curr = root
        while k > 0:
            if curr is not None:
                stack.append(curr)
                curr = curr.left
            elif stack:
                k -= 1
                curr = stack.pop()
                if k == 0:
                    break
                
                curr = curr.right
                
        return curr.val
        

File: LeetCode/Medium/0230-kth-smallest-element-in-a-bst/0230-kth-smallest-element-in-a-bst.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int kthSmallest(TreeNode* root, int k) {
        TreeNode* curr = root;
        stack<TreeNode*> stack;
        while (k > 0) {
            if (curr != nullptr) {
                stack.push(curr);
                curr = curr->left;
            } else if (!stack.empty()) {
                curr = stack.top();
                stack.pop();
                --k;
                if (k == 0) {
                    return curr->val;
                } else {
                    curr = curr->right;
                }
            }
        }
        
        return curr->val;
    }
};